CODEOWNERS

```diff
@@ -331,6 +331,12 @@ extensions/filters/http/oauth2 @derekargueta @mattklein123
 
 /*/extensions/health_checkers/common @zuercher @botengyao
 
+# HTTP credential injector
+/*/extensions/filters/http/credential_injector @zhaohuabing @kyessenov
+/*/extensions/credentials/common @zhaohuabing @kyessenov
+/*/extensions/credentials/basic_auth @zhaohuabing @kyessenov
+/*/extensions/credentials/bearer_token @zhaohuabing @kyessenov
```

Comment:
Looking at the implementation of these two injector and they are pretty similar. We should be able to refactor them into one similar implementation.

OTOH, what do you think even combine them into one injector, just taking whatever delivered via SDS into the header? Adding `Basic ` or `Bearer ` prefix and encoding user/password can be done at control plane.

Reply:
> But other credential types will need more complex processing, like interaction with third parties to get credentials, and I prefer to keep that logic inside the filter code, without involving an additional component in the control plane.

Yes for other credentials for sure, including getting tokens from cloud providers, oauth2 and oauth2 token exchange. I'm only suggesting a dumb header injector because that is more future proof and has some flexibility, for exampble api.github.com allows "Authorization: token <personal-access-token>" for authentication. And some HTTP server use basic without ':' separator but just a password.

Reply:
> > But other credential types will need more complex processing, like interaction with third parties to get credentials, and I prefer to keep that logic inside the filter code, without involving an additional component in the control plane.
> 
> Yes for other credentials for sure, including getting tokens from cloud providers, oauth2 and oauth2 token exchange. I'm only suggesting a dumb header injector because that is more future proof and has some flexibility, for exampble api.github.com allows "Authorization: token " for authentication. And some HTTP server use basic without ':' separator but just a password.

This filter is intended to allow extensions for all kinds of credentials, not just a dumb header injector. If people want that, they can use the existing header mutation capability or propose another filter in case sds is needed.

Reply:
Or maybe the "dumb header injector" can be one of the extension types of this filter, if someone really needs it.

Reply:
I'm on fence here while I understand that semantically this could be separate extensions, our [API style guide ](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md)says:

> The data plane APIs are primarily intended for machine generation and consumption. It is expected that the management server is responsible for mapping higher level configuration concepts to concrete API concepts. Similarly, static configuration fragments may be generated by tools and UIs, etc. The APIs and tools used to generate xDS configuration are beyond the scope of the definitions in this repository.

cc @envoyproxy/api-shepherds to weigh in.

Reply:
As a credential injector, I prefer to keep this an independent filter that get various credential from control-plane by `sds` and them inject the credential to the request headers with specified format.

And if some complex interactions/processing are necessary, IMO, control plane may have more advantages to enable credential sharing or caching.

Reply:
cc @zhaohuabing Seems you have updated some API and @lizan do some new comments. And I pretty like the idea to combine these extensions to one injector, esp considering that we have sds.

Reply:
> I'm on fence here while I understand that semantically this could be separate extensions, our [API style guide ](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md)says:
> 
> > The data plane APIs are primarily intended for machine generation and consumption. It is expected that the management server is responsible for mapping higher level configuration concepts to concrete API concepts. Similarly, static configuration fragments may be generated by tools and UIs, etc. The APIs and tools used to generate xDS configuration are beyond the scope of the definitions in this repository.
> 
> cc @envoyproxy/api-shepherds to weigh in.



> I'm on fence here while I understand that semantically this could be separate extensions, our [API style guide ](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md)says:
> 
> > The data plane APIs are primarily intended for machine generation and consumption. It is expected that the management server is responsible for mapping higher level configuration concepts to concrete API concepts. Similarly, static configuration fragments may be generated by tools and UIs, etc. The APIs and tools used to generate xDS configuration are beyond the scope of the definitions in this repository.
> 
> cc @envoyproxy/api-shepherds to weigh in.

If the configuration are generated by a control plane, I agree it m akes sense to combine the generic, baisc auth, and bearer token extensions. 

But I would insit that oauth2 client grant should be handled inside envoy in a separate extension.

Reply:
I'm on the fence as well. The intent of xDS was to act almost like an assembler layer. The reality is it is used by humans directly enough that this is probably wishful thinking. In any case, for this specific case, I'd defer to others.

Reply:
> At least I would insit that oauth2 client grant should be handled inside envoy in a separate extension. Otherwise, oauth2 client grant can't work without control plane.

Yes this is what I would insist too. For those need interact (e.g. oauth) / do some compute based on request (this could be potentially signing a JWT with the request headers etc), those need to be a separate extension and that's the purpose of credential injector.

> If the configuration are generated by a control plane, I agree it makes sense to combine the generic, baisc auth, and bearer token extensions. But I would argue that a clear, unambiguous API is better for users and the programmer of the control plane to understand and use. Both options work for the Control planes, and the code doesn't really care about it's three or one API that they're talking with.

I suggest we clearly documented those use case in the generic extension (or even at the credential injector filter itself) with concrete example. So folks know how to configure it for basic auth or bearer token. WDYT?

Reply:
Fair enough. I'll remove the basic auth and bearer token and document these two in the generic extension.